File: java-study/jvm-study/src/main/java/com/shanjiancaofu/jvmstudy/FinalizeEscapeGC.java
package com.shanjiancaofu.jvmstudy;

class FinalizeEscapeGC{
	public static  FinalizeEscapeGC fgc = null;
	public void isAlive(){
		System.out.println("i am alive");
	}
	@Override
	protected void finalize() throws Throwable {
		super.finalize();
		System.out.println("finalize method executed");
		// 重新赋值逃逸
		fgc = this;
	}

	public static void main(String[] args) {
		fgc = new FinalizeEscapeGC();
		fgc = null;
		System.gc();
		try {
			Thread.sleep(500);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		if(fgc != null){
			fgc.isAlive();
		}else{
			System.out.println("i am dead");
		}
		// 相同的代码， 逃逸失败
		fgc = null;
		System.gc();
		try {
			Thread.sleep(500);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
		if(fgc != null){
			fgc.isAlive();
		}else{
			System.out.println("i am dead");
		}
	}
}

File: java-study/scala-study/src/main/scala/com.shanjiancaofu.scala/helloworld/FinalizeEscapeGC.scala
package com.shanjiancaofu.scala.helloworld

class FinalizeEscapeGC(){
    var fgc : FinalizeEscapeGC = _
    def isAlive :Unit= {
        println("yes I'm alive")
    }
    override def finalize(): Unit = {
        super.finalize()
        println("finalize method executed！")

    }
}

object FinalizeEscapeGC {
    def main(args: Array[String]): Unit = {
       var  finalizeEscapeGC = new FinalizeEscapeGC
        var finalizeEscapeGC2 = new FinalizeEscapeGC
        finalizeEscapeGC.fgc = finalizeEscapeGC2
        finalizeEscapeGC.fgc =null
        System.gc()
        Thread.sleep(500)
        if(finalizeEscapeGC.fgc != null){
            finalizeEscapeGC.fgc.isAlive
        }else{
            println("i am dead")
        }
        // 相同的代码， 逃逸失败
        finalizeEscapeGC.fgc =null
        System.gc()
        Thread.sleep(500)
        if(finalizeEscapeGC.fgc != null){
            finalizeEscapeGC.fgc.isAlive
        }else{
            println("i am dead")
        }
    }
}
